### ISSUE TITLE
Handle OCI_SUCCESS_WITH_INFO

### ISSUE THREAD

**Issue by kyudin** [SOCI/soci#473]
Greetings. Why soci oracle backend does not handle OCI_SUCCESS_WITH_INFO result?
Just throw it as real error, but not as "Unknown error code" will help much.

```
    case OCI_ERROR:
    case OCI_SUCCESS_WITH_INFO:
        OCIErrorGet(errhp, 1, 0, &errcode,
             errbuf, sizeof(errbuf), OCI_HTYPE_ERROR);
        msg = reinterpret_cast<char*>(errbuf);
        errNum = static_cast<int>(errcode);
        break;
```

Actually probably it will be better to not treat OCI_SUCCESS_WITH_INFO as an exception, since actually it is not error, but just warning.
Thanks.

**Comment by vadz** [comment]
I see `OCI_SUCCESS_WITH_INFO` already being treated specially in src/backends/oracle/statement.cpp, so I'm not sure where exactly can we get "Unknown error" for it. I don't know much about OCI anyhow and so probably won't be able to do anything about it anyhow, but if someone is going to look at this, it would be certainly helpful to have a simple way of reproducing the problem, could you please provide one? Perhaps even as a (currently failing) new test in the test suite?

**Comment by kyudin** [comment]
vadz, actually `OCI_SUCCESS_WITH_INFO` can happen in any OCI call. For example yesterday we have this problem in production system, when I correct code for check what error actually was, we see, that info was: password will be expire in 10 days, that was on session open.

**Comment by snikulov** [comment]
@kyudin could you please provide information, which version/branch you using?

**Comment by kyudin** [comment]
@snikulov We use release 3.2.

**Comment by snikulov** [comment]
@vadz I see what @kyudin mean.
Every condition check ``res != OCI_SUCCESS`` will produce ``throw soci_error(...)``
Which, perhaps, not good in case when res equals OCI_SUCCESS_WITH_INFO...

**Comment by vadz** [comment]
Yes, I understand the problem, I'm still not sure about the fix though. If you look at `oracle/statement.cpp` you can see that the code already checks for `OCI_SUCCESS_WITH_INFO` in some places (`execute()` and `fetch()`), but not the others. So the question is whether we should also check for it everywhere (in which case it would probably be better to rename `throw_oracle_soci_error()` to something like `throw_oracle_soci_error_if_necessary()` and put the check inside it), or whether in some cases `OCI_SUCCESS_WITH_INFO` still needs to be treated as error? Someone who knows more about Oracle than me should decide...

Also, in the places where we do handle `OCI_SUCCESS_WITH_INFO` currently, we completely ignore it, which doesn't seem right neither, we need to have some way to return the message associated with the extra "info" to the user code and this is not specific to Oracle.

**Comment by mloskot** [comment]
My opinion is neutral.
AFAIK, @msobczak is the main author of Oracle support in SOCI and, since Maciej has lately re-activated in that area, I'd prefer to wait for his opinion.

**Comment by kyudin** [comment]
Probably just throw error with real information, but not with 'Unknown error', that will also helpfull.

**Comment by ddowling** [comment]
This issue just caught me on a production box. Connections were failing in OCISessionBegin() with a "ORA-28002: the password will expire within 7 days" warning. However as the soci::details::oracle::get_error_details() method did not handle the OCI_SUCCESS_WITH_INFO case then the reported error was the not very helpful "soci error: Unknown error code" exception. 

A partial fix for this is to just add "case OCI_SUCCESS_WITH_INFO:" after the "case OCI_ERROR:" line in the get_error_details method so at least a sensible message is reported in the exception that points the user in the correct direction.

**Comment by vadz** [comment]
@ddowling Could you please make a PR with the proposed change? I'd be glad to merge it, but I really don't have time for SOCI myself currently, sorry.

**Comment by vadz** [comment]
Fixed by #536.
